plugins/serialization/src/main/java/org/qbicc/plugin/serialization/HeapOutputStream.java

```diff
@@ -0,0 +1,98 @@
+package org.qbicc.plugin.serialization;
+
+import org.qbicc.context.CompilationContext;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.IdentityHashMap;
+
+class HeapOutputStream {
+    private static final int CHUNK_SIZE = 16; // 128 * 1024;  TODO: Set very small to test growing logic...don't leave it like this for long!
+
+    private final ByteOrder endianness;
+    private final IdentityHashMap<Object, Integer> objects = new IdentityHashMap<>();
+    private int lastIndex = 0;
+    private byte[] buffer;
+    private ByteBuffer out;
+
+    HeapOutputStream(ByteOrder endianness) {
+        this.endianness = endianness;
+        buffer = new byte[CHUNK_SIZE];
+        out = ByteBuffer.wrap(buffer).order(endianness);
+    }
+
+    private void ensureCapacity(int desired) {
+        if (out.position() + desired >= out.limit()) {
+            byte[] newBuf = new byte[buffer.length + CHUNK_SIZE];
+            System.arraycopy(buffer, 0, newBuf, 0, out.position());
```

Comment:
This could be simplified to `byte[] newBuf = Arrays.copyOf(buffer, buffer.length + CHUNK_SIZE);`.

Reply:
done